include/libopencm3/sam/common/pmc_common_all.h

```diff
@@ -27,160 +34,223 @@
 
 /* --- Power Management Controller (PMC) registers ----------------------- */
 
-/* System Clock Enable Register */
+/** System Clock Enable Register */
 #define PMC_SCER			MMIO32(PMC_BASE + 0x0000)
 
-/* System Clock Disable Register */
+/** System Clock Disable Register */
 #define PMC_SCDR			MMIO32(PMC_BASE + 0x0004)
 
-/* System Clock Status Register */
+/** System Clock Status Register */
 #define PMC_SCSR			MMIO32(PMC_BASE + 0x0008)
 
-/* Main Oscillator Register */
+/** Main Oscillator Register */
 #define CKGR_MOR			MMIO32(PMC_BASE + 0x0020)
 
-/* Main Clock Frequency Register */
+/** Main Clock Frequency Register */
 #define CKGR_MCFR			MMIO32(PMC_BASE + 0x0024)
 
-/* PLLA Register */
+/** PLLA Register */
 #define CKGR_PLLAR			MMIO32(PMC_BASE + 0x0028)
 
-/* Master Clock Register */
+/** Master Clock Register */
 #define PMC_MCKR			MMIO32(PMC_BASE + 0x0030)
 
-/* Programmable Clock 0 Register */
+/** Programmable Clock 0 Register */
 #define PMC_PCK0			MMIO32(PMC_BASE + 0x0040)
 
-/* Programmable Clock 1 Register */
+/** Programmable Clock 1 Register */
 #define PMC_PCK1			MMIO32(PMC_BASE + 0x0044)
 
-/* Programmable Clock 2 Register */
+/** Programmable Clock 2 Register */
 #define PMC_PCK2			MMIO32(PMC_BASE + 0x0048)
 
-/* Interrupt Enable Register */
+/** Interrupt Enable Register */
 #define PMC_IER				MMIO32(PMC_BASE + 0x0060)
 
-/* Interrupt Disable Register */
+/** Interrupt Disable Register */
 #define PMC_IDR				MMIO32(PMC_BASE + 0x0064)
 
-/* Status Register */
+/** Status Register */
 #define PMC_SR				MMIO32(PMC_BASE + 0x0068)
 
-/* Interrupt Mask Register */
+/** Interrupt Mask Register */
 #define PMC_IMR				MMIO32(PMC_BASE + 0x006C)
 
-/* Fast Startup Mode Register */
+/** Fast Startup Mode Register */
 #define PMC_FSMR			MMIO32(PMC_BASE + 0x0070)
 
-/* Fast Startup Polarity Register */
+/** Fast Startup Polarity Register */
 #define PMC_FSPR			MMIO32(PMC_BASE + 0x0074)
 
-/* Fault Output Clear Register */
+/** Fault Output Clear Register */
 #define PMC_FOCR			MMIO32(PMC_BASE + 0x0078)
 
-/* Write Protect Mode Register */
+/** Write Protect Mode Register */
 #define PMC_WPMR			MMIO32(PMC_BASE + 0x00E4)
 
-/* Write Protect Status Register */
+/** Write Protect Status Register */
 #define PMC_WPSR			MMIO32(PMC_BASE + 0x00E8)
 
 
 /* --- Register contents --------------------------------------------------- */
 
+/**
+ * @addtogroup pmc_scer PMC System Clock Enable Register (PMC_SCER)
+ * @{
+ */
 
-/* --- PMC System Clock Enable Register (PMC_SCER) ------------------------- */
-
-/* Programmable Clock Output Enable */
+/**
+ * Programmable Clock Output Enable
+ * @{
+ */
 #define PMC_SCER_PCK0			(0x01 << 8)
 #define PMC_SCER_PCK1			(0x01 << 9)
 #define PMC_SCER_PCK2			(0x01 << 10)
+/** @} */
```

Comment:
This makes problems right now.
The method with `@{` and `@}` works in general, but not very nice for defines.
With this code, there is no grouping, only the first define gets the comment.
Its not very bad, the other two defines appear alphabetically immediately below, so one doesn’t need to search far away.

I am not sure what to do exactly here.
My referenced example (nrf/common/i2c.h) does not have any comments on this low level, the defines just appear in alphabetical order on the documentation page.
Do you think I should just be happy with the alphabetical order?

Reply:
you need to eithter @defgroup or @addtogroup, if you want groups.  the @{ and @} won't work well by themselves.  the warning logs are normally useful.  Look at places like stm32, rather than nrf, it's had a lot more work on it over the years.

At this level, we would normally have something like : https://github.com/libopencm3/libopencm3/blob/master/include/libopencm3/stm32/common/adc_common_v2.h#L66-L83  with a group for all the values of a register, and that register in a group for the peripheral.  But you can also just go in steps, it doesn't have to do everything in one go.

Reply:
> with a group for all the values of a register, and that register in a group for the peripheral.

Yeah, that example looks good. I want to try that today. :)

Reply:
Done.

It turns out that NRF and STM32 do not differ much in this point, and I was pretty close to this:

> with a group for all the values of a register, and that register in a group for the peripheral.

It is just a bit more verbose now:
the precise answer for the following is to wrap related defines in @defgroup.

> The method with `@{` and `@}` works in general, but not very nice for defines.

---

I changed several lines to @defgroup, where I used @addtogroup before.
NRF seems to have been a bit less strict, but the decision for @defgroup and @addtogroup does not turn out to be difficult.
It just follows the place where the register address is defined: there the register is introduced, therefore the register contents group is defined with @defgroup.